rosbag2_transport/test/rosbag2_transport/test_play_loop.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+// Copyright 2020, TNG Technology Consulting GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <chrono>
+#include <future>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "rosbag2_test_common/subscription_manager.hpp"
+
+#include "rosbag2_transport/rosbag2_transport.hpp"
+
+#include "test_msgs/msg/basic_types.hpp"
+#include "test_msgs/message_fixtures.hpp"
+
+#include "rosbag2_play_test_fixture.hpp"
+
+TEST_F(RosBag2PlayTestFixture, messages_played_in_loop) {
+  auto primitive_message1 = get_messages_basic_types()[0];
+  const int test_value = 42;
+  primitive_message1->int32_value = test_value;
+
+  auto topic_types = std::vector<rosbag2_storage::TopicMetadata>{
+    {"loop_test_topic", "test_msgs/BasicTypes", "", ""}
+  };
+
+  size_t num_messages = 3;
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> messages(num_messages,
+    serialize_test_message("loop_test_topic", 700, primitive_message1));
+
+  auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+  prepared_mock_reader->prepare(messages, topic_types);
+  reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+
+  size_t expected_number_of_messages = num_messages * 2;
+  sub_->add_subscription<test_msgs::msg::BasicTypes>(
+    "/loop_test_topic",
+    expected_number_of_messages);
+
+  auto await_received_messages = sub_->spin_subscriptions();
+
+  auto rosbag2_transport_ptr = std::make_shared<rosbag2_transport::Rosbag2Transport>(
+    reader_,
+    writer_,
+    info_);
+  bool loop_playback = true;
+  std::thread loop_thread(&rosbag2_transport::Rosbag2Transport::play, rosbag2_transport_ptr,
+    storage_options_,
+    rosbag2_transport::PlayOptions{1000, "", 1.0, {}, {}, loop_playback});
+  std::this_thread::sleep_for(std::chrono_literals::operator""s(1));
```

Comment:
Why are we sleeping here? Is this robust to slow machines?

Additionally, instead of trying to do a literal, you can supply `std::chrono::seconds(1)` - which should fix the OSX build

Reply:
I believe that should also fix the windows warnings

Reply:
Yes, I actually added the sleep_for to make sure that enough messages are played for the test to pass, but if you think it wouldn't be necessary I can remove it

Reply:
I think that there are better ways. Sleep-based tests are inherently flaky and prone to randomly breaking CI. You'd be much better off waiting for a specific condition to be met (up to a reasonable timeout)

Reply:
See the [`spin_and_wait_for` utility](https://github.com/ros2/rosbag2/blob/master/rosbag2_test_common/include/rosbag2_test_common/wait_for.hpp#L25) and [an example of its usage](https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/test/rosbag2_transport/test_record.cpp#L126)

This way, rather than waiting a fixed amount of time, your test can finish as soon as the condition is met, and time eventually if the condition is never met.

Reply:
I changed this a little bit, I noticed that the `SubscriptionManager` returns a future and when we call `await_received_messages` it will automatically wait until the messages are received, and this should be equivalent with what you suggested.